clconf/template.go

```diff
@@ -0,0 +1,221 @@
+package clconf
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall"
+)
+
+const setuid = 4
+const setgid = 2
+const sticky = 1
+
+// TemplateOptions are settings for ProcessTemplates.
+type TemplateOptions struct {
+	// CopyTemplatePerms uses the existing template permissions instead of FileMode for template
+	// permissions
+	CopyTemplatePerms bool
+	// Flatten flattens the templates into the root of the dest instead of the preserving
+	// the relative path under the source.
+	Flatten bool
+	// KeepExistingPerms determines whether to use existing permissions on template files that are overwritten.
+	KeepExistingPerms bool
+	// Rm determines whether template files distinct from their target are deleted after processing.
+	Rm bool
+	// FileMode is the permissions to apply to template files when writing.
+	FileMode os.FileMode
+	// DirMode is the permission similar to FileMode but for new folders.
+	DirMode os.FileMode
+	// Extension is the extension to use when searching folders. If missing all files will be used.
+	// The extension is stripped from the file name when templating.
+	Extension string
+}
+
+type pathWithRelative struct {
+	full string
+	rel  string
+}
+
+// TemplateResult stores the result of a single template processing.
+type TemplateResult struct {
+	Src  string
+	Dest string
+}
+
+// ProcessTemplates processes templates. If dest is non empty it must be a folder into which
+// templates will be placed after processing (the folder will be created if necessary). If empty
```

Comment:
It has to be a folder?  I thought the design goal was to behave like `cp`... With `cp` you can have the destination of a `file` which allows you to _change the output name_.  Is that not a desired feature here?

Reply:
Yes, it must be a folder in the current code. We could add another function that takes a single source and single destination, but that would likely be called `ProcessTemplate`, so wouldn't impact this API later.